### PR TITLE
feat: add rsagen codegen

### PR DESCRIPTION
**Pull request by tie**
This PR adds rsagen command that generates Go source code for crypto/rsa.PublicKey variables from PEM-encoded RSA public keys. The command generates a slice of all input public keys or, if in single mode, expects one RSA public key and outputs single rsa.PublicKey variable.

Also adds testscript dependency we can use to test other commands.

<details>
<summary>Example</summary>

__Run__
```
keygen -single -f input.pem -o output.go -pkg seed -var _PublicKeyDNS
```

__Input__
```
-----BEGIN RSA PUBLIC KEY-----
MIIBCgKCAQEAyr+18Rex2ohtVy8sroGPBwXD3DOoKCSpjDqYoXgCqB7ioln4eDCF
fOBUlfXUEvM/fnKCpF46VkAftlb4VuPDeQSS/ZxZYEGqHaywlroVnXHIjgqoxiAd
192xRGreuXIaUKmkwlM9JID9WS2jUsTpzQ91L8MEPLJ/4zrBwZua8W5fECwCCh2c
9G5IzzBm+otMS/YKwmR1olzRCyEkyAEjXWqBI9Ftv5eG8m0VkBzOG655WIYdyV0H
fDK/NWcvGqa0w/nriMD6mDjKOryamw0OP9QuYgMN0C9xMW9y8SmP4h92OAWodTYg
Y1hZCxdv6cs5UnW9+PWvS+WIbkh+GaWYxwIDAQAB
-----END RSA PUBLIC KEY-----
```

__Output__
```
// Code generated by rsagen, DO NOT EDIT.
//
// -----BEGIN RSA PUBLIC KEY-----
// MIIBCgKCAQEAyr+18Rex2ohtVy8sroGPBwXD3DOoKCSpjDqYoXgCqB7ioln4eDCF
// fOBUlfXUEvM/fnKCpF46VkAftlb4VuPDeQSS/ZxZYEGqHaywlroVnXHIjgqoxiAd
// 192xRGreuXIaUKmkwlM9JID9WS2jUsTpzQ91L8MEPLJ/4zrBwZua8W5fECwCCh2c
// 9G5IzzBm+otMS/YKwmR1olzRCyEkyAEjXWqBI9Ftv5eG8m0VkBzOG655WIYdyV0H
// fDK/NWcvGqa0w/nriMD6mDjKOryamw0OP9QuYgMN0C9xMW9y8SmP4h92OAWodTYg
// Y1hZCxdv6cs5UnW9+PWvS+WIbkh+GaWYxwIDAQAB
// -----END RSA PUBLIC KEY-----
//

package seeds

import (
	"crypto/rsa"
	"math/big"
)

var _ = (*big.Int)(nil)

var _PublicKeyDNS = &rsa.PublicKey{
	// Fingerprint: 6f3a701151477715
	E: 65537,
	N: new(big.Int).SetBytes([]byte{
		0xca, 0xbf, 0xb5, 0xf1, 0x17, 0xb1, 0xda, 0x88,
		0x6d, 0x57, 0x2f, 0x2c, 0xae, 0x81, 0x8f, 0x07,
		0x05, 0xc3, 0xdc, 0x33, 0xa8, 0x28, 0x24, 0xa9,
		0x8c, 0x3a, 0x98, 0xa1, 0x78, 0x02, 0xa8, 0x1e,
		0xe2, 0xa2, 0x59, 0xf8, 0x78, 0x30, 0x85, 0x7c,
		0xe0, 0x54, 0x95, 0xf5, 0xd4, 0x12, 0xf3, 0x3f,
		0x7e, 0x72, 0x82, 0xa4, 0x5e, 0x3a, 0x56, 0x40,
		0x1f, 0xb6, 0x56, 0xf8, 0x56, 0xe3, 0xc3, 0x79,
		0x04, 0x92, 0xfd, 0x9c, 0x59, 0x60, 0x41, 0xaa,
		0x1d, 0xac, 0xb0, 0x96, 0xba, 0x15, 0x9d, 0x71,
		0xc8, 0x8e, 0x0a, 0xa8, 0xc6, 0x20, 0x1d, 0xd7,
		0xdd, 0xb1, 0x44, 0x6a, 0xde, 0xb9, 0x72, 0x1a,
		0x50, 0xa9, 0xa4, 0xc2, 0x53, 0x3d, 0x24, 0x80,
		0xfd, 0x59, 0x2d, 0xa3, 0x52, 0xc4, 0xe9, 0xcd,
		0x0f, 0x75, 0x2f, 0xc3, 0x04, 0x3c, 0xb2, 0x7f,
		0xe3, 0x3a, 0xc1, 0xc1, 0x9b, 0x9a, 0xf1, 0x6e,
		0x5f, 0x10, 0x2c, 0x02, 0x0a, 0x1d, 0x9c, 0xf4,
		0x6e, 0x48, 0xcf, 0x30, 0x66, 0xfa, 0x8b, 0x4c,
		0x4b, 0xf6, 0x0a, 0xc2, 0x64, 0x75, 0xa2, 0x5c,
		0xd1, 0x0b, 0x21, 0x24, 0xc8, 0x01, 0x23, 0x5d,
		0x6a, 0x81, 0x23, 0xd1, 0x6d, 0xbf, 0x97, 0x86,
		0xf2, 0x6d, 0x15, 0x90, 0x1c, 0xce, 0x1b, 0xae,
		0x79, 0x58, 0x86, 0x1d, 0xc9, 0x5d, 0x07, 0x7c,
		0x32, 0xbf, 0x35, 0x67, 0x2f, 0x1a, 0xa6, 0xb4,
		0xc3, 0xf9, 0xeb, 0x88, 0xc0, 0xfa, 0x98, 0x38,
		0xca, 0x3a, 0xbc, 0x9a, 0x9b, 0x0d, 0x0e, 0x3f,
		0xd4, 0x2e, 0x62, 0x03, 0x0d, 0xd0, 0x2f, 0x71,
		0x31, 0x6f, 0x72, 0xf1, 0x29, 0x8f, 0xe2, 0x1f,
		0x76, 0x38, 0x05, 0xa8, 0x75, 0x36, 0x20, 0x63,
		0x58, 0x59, 0x0b, 0x17, 0x6f, 0xe9, 0xcb, 0x39,
		0x52, 0x75, 0xbd, 0xf8, 0xf5, 0xaf, 0x4b, 0xe5,
		0x88, 0x6e, 0x48, 0x7e, 0x19, 0xa5, 0x98, 0xc7,
	}),
}
```
</details>

<details>
<summary>Test example</summary>

```
// Code generated by rsagen, DO NOT EDIT.

package main

import (
	"fmt"
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"

	"github.com/gotd/td/internal/crypto"
)

func TestFingerprint(t *testing.T) {
	fingerprint := func(pubkey *rsa.PublicKey) string {
		return fmt.Sprintf("%08x", uint64(crypto.RSAFingerprint(pubkey)))
	}
	expected := []string{
		"c3b42b026ce86b21",
		"bc35f3509f7b7a5",
		"15ae5fa8b5529542",
		"aeae98e13cd7f94f",
		"5a181b2235057d98",
	}

	require.Len(t, PK, len(fingerprints))
	for i, pubkey := range PK {
		assert.Equal(t, expected[i], fingerprint(pubkey))
	}
}
```
</details>

### CODE REVIEW

File: cmd/rsagen/main.go [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+		}
+		return keys[0], nil
+	},
+}
+
+func main() {
+	os.Exit(main1())
+}
+
+func main1() int {
```
Comment (ernado) [inline cmd/rsagen/main.go:51]: please use 

```go
func run() error {
// ...
}
```
pattern

File: cmd/rsagen/main.go [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+		inPath       = ""
+		outPath      = ""
+		tplPath      = ""
+		tplName      = "main.tmpl"
+		pkgName      = "main"
+		varName      = "PK"
+		singleMode   = false
+		formatOutput = true
+	)
+	flag.StringVar(
```
Comment (ernado) [inline cmd/rsagen/main.go:63]: Please use (at least for consistency with other commands):
```
var (
    inPath = flag.String() // ...
)
```
The `StringVar` patter is too verbose.

File: cmd/rsagen/main.go [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+		&inPath, "f", inPath,
+		"input path (defaults to stdin)",
+	)
+	flag.StringVar(
+		&outPath, "o", outPath,
+		"output path (defaults to stdout)",
+	)
+	flag.StringVar(
```
Comment (ernado) [inline cmd/rsagen/main.go:71]: I'm not sure that so much flexibility is needed, probably it is excessive to expose template dir and name.

**Review by ernado — CHANGES_REQUESTED** [review]
Thank you for review, I've added some comments.

Anyway I'm not sure that we need such codegen at all, parsing keys in runtime should be OK and it easier to verify (because they are PEM-encoded) than raw `[]byte` sequences from bigint.

So if there is no hard reason to introduce rsagen, I would stick to runtime parsing.

File: cmd/rsagen/main.go (hunk 3 above)
Comment (tie) [reply cmd/rsagen/main.go]: Perhaps you are right, though it was quite useful when I wanted to experiment with multiple different templates.

Also, with Go’s `fs.FS` it’s a matter of a few flags and a simple conditional branch.
```
if *tplPath != "" {
	fsys = os.DirFS(*tplPath)
}
```

So there is almost no cost in keeping this extra feature.

**Comment by tie** [comment]
I think I’ve addressed your concert about verification in the last commits.
1. The rsagen command supports generating tests that verify the key’s fingerprint.
2. File’s header comment now contains PEM-encoded public keys.

File: cmd/rsagen/main.go (hunk 3 above)
Comment (tie) [reply cmd/rsagen/main.go]: The last revision ended up using template name argument for switching to tests codegen. Adding template directory is a small bonus on top (just three lines of code above, shouldn’t be a problem). Therefore closing as resolved.

**Review by ernado — APPROVED** [review]

**Pull request merged**